petclock.asm

```diff
@@ -851,7 +920,68 @@ UpdateJiffyClock:
                 dex                     ; Decrease minute count and continue if necessary
                 bne @mmloop
 
-@addjiffy:      lda remainder           ; Check if we have anything to add to current
+@addseconds:    
+                
+.if C64         ; On the C64, use the seconds and tenths in our clock structure
+                ldx SecTens
+                ldy SecDigits
+                jsr GetCharsValue       ; Convert second characters to value
+
+                cpx #0                  ; No seconds? Skip adding.
+                beq @addtenths
+
+                ldy #SECOND_JIFFIES
+                jsr Multiply
+
+                clc
+                lda resultLo            ; Add result and extend for carry
+                adc remainder
+                sta remainder
+                lda resultHi
+                adc remainder+1
+                sta remainder+1
+                lda #0
+                adc remainder+2
+                sta remainder+2
+
+@addtenths:     lda Tenths
+                cmp #'0'
+                beq @writejiffy
+
+                sec
+                sbc #'0'
+                tax
+                ldy #SECOND_JIFFIES/10
+                jsr Multiply
+
+                clc
+                lda resultLo
+                adc remainder
+                sta remainder
+                lda #0
+                adc remainder+1
+                sta remainder+1
+                lda #0
+                adc remainder+2
+                sta remainder+2
+
+@writejiffy:    lda remainder           ; Load jiffies in registers, so we can keep 
+                ldx remainder+1         ;   the interrupts disabled for the shortest 
+                ldy remainder+2         ;   duration possible.
+
+                sei

```

Comment:
Random Q: are you worried about timer interrupts coming in or other IRQs?  If timer, they are normally NMIs and this wouldn't stop them.  But raster interrupts and so on it does mask.   I could be way off base but I thought I remember the 6526's TOD being wired to the NMI.

Reply:
I'm really only worried about one IRQ, being the one that updates the jiffy timer bytes. It is a "regular" software IRQ that does this, which is therefore stopped by the SEI.
Actually, after writing this code for the PET, I found out that the RDTIM/SETTIM kernal code on the C64 uses exactly the same approach.

If the 6526 TOD is wired to the NMI is irrelevant in this case, because; 
1. this is PET only code, so there are no 6526's to worry about, and
2. the code interacts with the jiffy timer, not the 6526 TOD. 

The somewhat surprising thing is that on the C64, the 6526 TOD and the jiffy timer are completely disconnected, unless you the programmer connect them yourselves. Which is what PET Clock does at program start and exit.
I've read speculation that this disconnect of timers is one of the consequences of Commodore rushing out the kernal for the C64.

Reply:
I just realized that point 1 in my previous comment was nonsense, as this is the code that sets the jiffy timer on program exit on the C64. The rest still stands though. :)